### PR TITLE
Implementing extensible enums as structs

### PR DESCRIPTION
**Pull request by dsgouda**
- Still need to update references to the modeler package, dependent on https://github.com/Azure/autorest.modeler/pull/27

### CODE REVIEW

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

File: src/azure/CodeGeneratorCsa.cs [hunk 1]
```diff
@@ -84,9 +84,22 @@ public class CodeGeneratorCsa : CodeGeneratorCs
             // Enums
             foreach (EnumTypeCs enumType in codeModel.EnumTypes)
             {
-                var enumTemplate = new EnumTemplate {Model = enumType};
-                await Write(enumTemplate,Path.Combine(Settings.Instance.ModelsName,
-                    $"{enumTemplate.Model.Name}{ImplementationFileExtension}"));
+                if((Settings.Instance.CustomSettings.ContainsKey("ExtensibleEnums") && (bool)Settings.Instance.CustomSettings["ExtensibleEnums"]))
```
Comment (olydis) [inline src/azure/CodeGeneratorCsa.cs]: use extension method `Get<bool>` for conciseness :)

File: src/vanilla/Model/CompositeTypeCs.cs [hunk 2]
```diff
@@ -159,13 +159,29 @@ public string Signature
...
+                        
+                        // we need some special signatures for extensible enums
+                        if(property.ModelType is EnumType && Settings.Instance.CustomSettings.ContainsKey("ExtensibleEnums") && (bool)Settings.Instance.CustomSettings["ExtensibleEnums"])
                         {
-                            format = "{0} {1} = " + property.DefaultValue;
+                            // for people who really want defaults to properties (be aware of the PATCH operation consequences!)
+                            var enumProp = property.ModelType as EnumType;
+                            if(property.UseDefaultInConstructor && enumProp.DefaultValue!=null)
+                            {
+                                format = "{0} {1} = " + enumProp.DefaultValue;
```
Comment (olydis) [inline src/vanilla/Model/CompositeTypeCs.cs]: why is this a special case? I could very well be missing something right now, but why would we take DefaultValue out of the enum *type* instead of the property?

File: test/autorest.csharp.test.csproj [hunk 3]
```diff
@@ -36,6 +36,7 @@
     <PackageReference Include="System.Reflection.TypeExtensions" Version="4.3.0" />
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.3.1" />
     <PackageReference Include="System.Net.Http" Version="4.3.2" />
+    <PackageReference Include="System.Collections.Concurrent" Version="4.3.0" />
```
Comment (olydis) [inline test/autorest.csharp.test.csproj]: maybe I missed latest developments, but is there still a collection required in the way we model enums now?

File: test/vanilla/Expected/AcceptanceTests/BodyString/IStringModel.cs [hunk 4]
```diff
@@ -91,8 +91,7 @@ public partial interface IStringModel
         Task<HttpOperationResponse> PutEmptyWithHttpMessagesAsync(string stringBody, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken));
         /// <summary>
         /// Get mbcs string value
-        /// '啊齄丂狛狜隣郎隣兀﨩ˊ▇█〞〡￤℡㈱‐ー﹡﹢﹫、〓ⅰⅹ⒈€㈠㈩ⅠⅫ！￣ぁんァヶΑ︴АЯаяāɡㄅㄩ─╋︵﹄︻︱︳︴ⅰⅹɑɡ〇〾⿻⺁䜣€
-        /// '
+        /// '啊齄丂狛狜隣郎隣兀﨩ˊ〞〡￤℡㈱‐ー﹡﹢﹫、〓ⅰⅹ⒈€㈠㈩ⅠⅫ！￣ぁんァヶΑ︴АЯаяāɡㄅㄩ─╋︵﹄︻︱︳︴ⅰⅹɑɡ〇〾⿻⺁䜣€'
```
Comment (olydis) [inline test/vanilla/Expected/AcceptanceTests/BodyString/IStringModel.cs]: where does this change come from? is this from bumping the reference to common?

File: test/autorest.csharp.test.csproj (hunk 3 above)
Comment (dsgouda) [reply test/autorest.csharp.test.csproj]: hmm... that's true, let me see if I can get rid of it

File: test/vanilla/Expected/AcceptanceTests/BodyString/IStringModel.cs (hunk 4 above)
Comment (dsgouda) [reply test/vanilla/Expected/AcceptanceTests/BodyString/IStringModel.cs]: Partly common/modeler and partly the testserver, I don't have control over it 😞

Comment (olydis) [reply test/vanilla/Expected/AcceptanceTests/BodyString/IStringModel.cs]: thought so, no problem just wanted to make sure ;)

File: src/vanilla/Model/CompositeTypeCs.cs (hunk 2 above)
Comment (dsgouda) [reply src/vanilla/Model/CompositeTypeCs.cs]: This is slightly different `Property.ModelType.DefaultValue`

Comment (olydis) [reply src/vanilla/Model/CompositeTypeCs.cs]: How? My point is: default values are something you specify on *usage* of a type (property, parameter), not the type itself - so why resort to the type here? In essence, it's a weird glitch in our current C# types that types even *have* a `DefaultValue`. In other words: If I have a Swagger with some extensible enum and say `default: 3`, are you sure that `enumProp.DefaultValue` will contain 3?

File: src/azure/CodeGeneratorCsa.cs (hunk 1 above)
Comment (dsgouda) [reply src/azure/CodeGeneratorCsa.cs]: I don't think those extensions methods are included in this project, sounds like too much work trying to fix that.

Comment (olydis) [reply src/azure/CodeGeneratorCsa.cs]: they are in common AFAIK

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

File: src/vanilla/Model/CompositeTypeCs.cs (hunk 2 above)
Comment (dsgouda) [reply src/vanilla/Model/CompositeTypeCs.cs]: I remember now, the thing is, by default, we set the type for the default property as `property.ModelTypeName` but since we are creating a new type for the enum, we need `enumProp.ClassName`.

Comment (olydis) [reply src/vanilla/Model/CompositeTypeCs.cs]: aha! but shouldn't that happen at the level of `property.DefaultValue`? Like this, this is very problematic:
What will `(property.ModelType as EnumType).DefaultValue` contain if one and the same enum is used twice, and with different default values? It simply *can't* be the correct value in both cases since `property.ModelType as EnumType` will point to the same object. *Only* `property.DefaultValue` has the fidelity to distinguish both usages. This is similar to the type definition vs. type usage issues we once had with `required` 🙂

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        /// </summary>
+        public static implicit operator DaysOfWeekExtensibleEnum(string value)
+        {
+            return new DaysOfWeekExtensibleEnum(value);
+        }
+
+        /// <summary>
+        /// Implicit operator to convert DaysOfWeekExtensibleEnum to string
+        /// </summary>
+        public static implicit operator string(DaysOfWeekExtensibleEnum e)
```
Comment (brjohnstmsft) [inline test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: This should probably be explicit, not implicit. Otherwise type safety is easily compromised, defeating the purpose of having this type in the first place.

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    /// </summary>
+    [JsonConverter(typeof(DaysOfWeekExtensibleEnumConverter))]
+    public struct DaysOfWeekExtensibleEnum : System.IEquatable<DaysOfWeekExtensibleEnum>
+    {
+        private DaysOfWeekExtensibleEnum(string underlyingValue)
+        {
+            UnderlyingValue=underlyingValue;
+        }
+
+        public static readonly DaysOfWeekExtensibleEnum Monday = "Monday";
```
Comment (brjohnstmsft) [inline test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: I have a feature request that will help us move the Azure Search SDK to use this new feature. Please generate a nested class to contain the string constants, and then refer to them here. 

For example, see the `AsString` class in this file: https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SDKs/Search/DataPlane/Microsoft.Azure.Search/Customizations/Indexes/Models/AnalyzerName.cs

The reason we need this is that sometimes we need to refer to one of the extensible enum values from the parameter to an attribute, and such parameters must be compile-time constants.

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/IntEnum.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+        /// <summary>
+        /// two
+        /// </summary>
+                public static readonly IntEnum Two = "2";
+
+        /// <summary>
+        /// three
+        /// </summary>
+                public static readonly IntEnum Three = "3";
```
Comment (brjohnstmsft) [inline test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/IntEnum.cs]: nit: Weird indentation

File: src/vanilla/Model/CompositeTypeCs.cs (hunk 2 above)
Comment (dsgouda) [reply src/vanilla/Model/CompositeTypeCs.cs]: Ahh okay, I see, I will use the defaultValue from property then, ( I still need the typecasting to enumtype for reasons mentioned above)

File: src/vanilla/Model/CompositeTypeCs.cs [hunk 8]
```diff
@@ -165,9 +165,9 @@ public string Signature
                         {
                             // for people who really want defaults to properties (be aware of the PATCH operation consequences!)
                             var enumProp = property.ModelType as EnumType;
-                            if(property.UseDefaultInConstructor && enumProp.DefaultValue!=null)
+                            if(property.UseDefaultInConstructor && property.DefaultValue!=null)
```
Comment (olydis) [inline src/vanilla/Model/CompositeTypeCs.cs]: weeeeell and now we can undo the branching here - note that also `property.ModelTypeName` should probably point to `enumProp.ClassName`, but not sure... as little code duplication as possible 😉

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs (hunk 6 above)
Comment (dsgouda) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: Seems doable and looks like something we can implement incrementally, will open an issue describing this feature requirement.

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs (hunk 5 above)
Comment (dsgouda) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: This was an ask from the dotnet sdk team; also this allows a natural transition for `modelAstString:true` in the future.  
@shahabhijeet @markcowl if you can verify this is okay, I can merge this ASAP

Comment (fearthecowboy) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: @brjohnstmsft  -- From my perspective, I would have went with the explicit operator, but I'm not super passionate one way or the other.  The SDK team did make the point that it was a no-op code change for those who had been using strings.

File: src/vanilla/Model/CompositeTypeCs.cs [hunk 9]
```diff
@@ -159,13 +159,20 @@ public string Signature
                     foreach (PropertyCs property in Parameters.Where(p => !p.UnderlyingProperty.IsConstant).Select(p => p.UnderlyingProperty))
                     {
                         string format = (property.IsRequired ? "{0} {1}" : "{0} {1} = default({0})");
+                        
+                        // we need some special signatures for extensible enums
+                        var modelTypeName = 
+                            (property.ModelType is EnumType && (property.ModelType as EnumType).ModelAsExtensible)?
+                                 (property.ModelType as EnumType).ClassName : property.ModelTypeName;
```
Comment (olydis) [inline src/vanilla/Model/CompositeTypeCs.cs]: just to raise the point again: why don't we make `ModelTypeName` return `ClassName` of the enum? Did you investigate that path? Like this, I'm just super afraid about all other locations using `property.ModelTypeName`! Why don't they need the enums class name?

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs (hunk 5 above)
Comment (brjohnstmsft) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: @fearthecowboy If folks don't care about type safety, then they can keep using strings. This is a new feature of AutoRest; IMO it's more than reasonable to require a few code changes to start using it, especially when it's as simple (and clear) as adding some explicit casts.

I'm curious to hear what @shahabhijeet and @markcowl have to say.

Comment (shahabhijeet) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: What I know from hallway conversation was that, implicit conversion requirements are coming from PowerShell perspective. Mark will be right person to comment on the necessity to have support for implicit conversion.

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs (hunk 5 above)
Comment (fearthecowboy) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: Perhaps we should make implicit/explicit optional.

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

File: test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs (hunk 5 above)
Comment (brjohnstmsft) [reply test/vanilla/Expected/AcceptanceTests/ExtensibleEnums/Models/DaysOfWeekExtensibleEnum.cs]: @fearthecowboy Sounds good to me.

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by dsgouda** [comment]
>start win32-x64

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 98% 

The following 7 features are not covered by your tests:
❌ [`HttpClientFailure407Delete`](https://github.com/Azure/autorest.testserver/search?q=HttpClientFailure407Delete)
❌ [`HttpRedirect301Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Put)
❌ [`HttpRedirect302Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Patch)
❌ [`HttpRedirect303Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect303Post)
❌ [`ResponsesScenarioNoModelEmptyBody`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioNoModelEmptyBody)
❌ [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated)
❌ [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated)


## Azure: 96% 

The following 4 features are not covered by your tests:
❌ [`postParameterGroupingMultipleParameterGroups`](https://github.com/Azure/autorest.testserver/search?q=postParameterGroupingMultipleParameterGroups)
❌ [`postParameterGroupingOptionalParameters`](https://github.com/Azure/autorest.testserver/search?q=postParameterGroupingOptionalParameters)
❌ [`postParameterGroupingRequiredParameters`](https://github.com/Azure/autorest.testserver/search?q=postParameterGroupingRequiredParameters)
❌ [`postParameterGroupingSharedParameterGroupObject`](https://github.com/Azure/autorest.testserver/search?q=postParameterGroupingSharedParameterGroupObject)

**Comment by olydis** [comment]
@dsgouda will verify again, but the Azure coverage problems were not happening for me before - just making sure we don't regress here

File: test/azure/AcceptanceTests.cs [hunk 10]
```diff
@@ -659,75 +657,5 @@ public void DurationTests()
             }
         }
 
-        [Fact]
-        public void ParameterGroupingTests()
```
Comment (olydis) [inline test/azure/AcceptanceTests.cs]: huh, why did you remove these?

Comment (dsgouda) [reply test/azure/AcceptanceTests.cs]: @olydis just pushed a change undoing this (which should fail on the CI), and I run into compilation issues, since you were working on this, can you take a look?

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 98% 

The following 7 features are not covered by your tests:
❌ [`HttpClientFailure407Delete`](https://github.com/Azure/autorest.testserver/search?q=HttpClientFailure407Delete)
❌ [`HttpRedirect301Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Put)
❌ [`HttpRedirect302Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Patch)
❌ [`HttpRedirect303Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect303Post)
❌ [`ResponsesScenarioNoModelEmptyBody`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioNoModelEmptyBody)
❌ [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated)
❌ [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated)


## Azure: 100% ✔️

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 98% 

The following 7 features are not covered by your tests:
❌ [`HttpClientFailure407Delete`](https://github.com/Azure/autorest.testserver/search?q=HttpClientFailure407Delete)
❌ [`HttpRedirect301Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Put)
❌ [`HttpRedirect302Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Patch)
❌ [`HttpRedirect303Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect303Post)
❌ [`ResponsesScenarioNoModelEmptyBody`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioNoModelEmptyBody)
❌ [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated)
❌ [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated)


## Azure: 100% ✔️

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~
